pkg/utils/interceptors/logging.go

```diff
@@ -0,0 +1,119 @@
+			Str("method", method).
+			Str("version", version).
+			Uint32("code", uint32(code)).
+			Dur("duration", duration).
+			Logger()
+
+		// Log based on the error code
+		switch code {
+		case codes.OK:
+			logctx.Info().Msg(info.FullMethod)
+		case codes.Unknown:
+			logctx.Error().Err(err).Msgf("unknown error handling %s", info.FullMethod)
+		case codes.Canceled, codes.InvalidArgument, codes.NotFound, codes.AlreadyExists, codes.Unauthenticated:
+			logctx.Info().Err(err).Msg(info.FullMethod)
+		case codes.DeadlineExceeded, codes.PermissionDenied, codes.ResourceExhausted, codes.FailedPrecondition, codes.Aborted, codes.OutOfRange, codes.Unavailable:
+			logctx.Warn().Err(err).Msg(info.FullMethod)
+		case codes.Unimplemented, codes.Internal, codes.DataLoss:
+			logctx.Error().Err(err).Str("full_method", info.FullMethod).Msg(err.Error())
+		default:
+			logctx.Error().Err(err).Msgf("unhandled error code %s: %s", code, info.FullMethod)
+		}
+
+		return rep, err
+	}
+}
+
+// StreamLogging handles generic logging of RPC methods to zerolog.
+func StreamLogging() grpc.StreamServerInterceptor {
+	version := pkg.Version()
+	return func(srv interface{}, stream grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) (err error) {
+		// Parse the method for tags
+		service, method := ParseMethod(info.FullMethod)
+
+		// Handle the request and trace how long the request takes.
+		start := time.Now()
+		err = handler(srv, stream)
+		duration := time.Since(start)
+		code := status.Code(err)
+
+		// Prepare log context for logging
+		logctx := log.With().
+			Str("type", "stream").
+			Str("stream_type", StreamType(info)).
+			Str("service", service).
+			Str("method", method).
+			Str("version", version).
+			Uint32("code", uint32(code)).
+			Dur("duration", duration).
+			Logger()
+
+		switch code {
+		case codes.OK:
+			logctx.Info().Msg(info.FullMethod)
+		case codes.Unknown:
+			logctx.Error().Err(err).Msgf("unknown error handling %s", info.FullMethod)
+		case codes.Canceled, codes.InvalidArgument, codes.NotFound, codes.AlreadyExists, codes.Unauthenticated:
+			logctx.Info().Err(err).Msg(info.FullMethod)
+		case codes.DeadlineExceeded, codes.PermissionDenied, codes.ResourceExhausted, codes.FailedPrecondition, codes.Aborted, codes.OutOfRange, codes.Unavailable:
+			logctx.Warn().Err(err).Msg(info.FullMethod)
+		case codes.Unimplemented, codes.Internal, codes.DataLoss:
+			logctx.Error().Err(err).Str("full_method", info.FullMethod).Msg(err.Error())
+		default:
+			logctx.Error().Err(err).Msgf("unhandled error code %s: %s", code, info.FullMethod)
+		}
+		return err
+	}
+}
+
+func ParseMethod(method string) (string, string) {
+	method = strings.TrimPrefix(method, "/") // remove leading slash
+	if i := strings.Index(method, "/"); i >= 0 {
+		return method[:i], method[i+1:]
+	}
+	return "unknown", "unknown"
+}
+
+func StreamType(info *grpc.StreamServerInfo) string {
+	if !info.IsClientStream && !info.IsServerStream {
```

Comment:
Since this is only called in `StreamLogging`, which only accepts `grpc.ServerStreams`, do we really need to check here for unary RPC's?

Reply:
I think we do - if we don't handle all cases of the different combination of bools then the default will be "bidirectional" and that's not the case when they're both false.